central/reports/snapshot/datastore/datastore_impl.go

```diff
@@ -0,0 +1,134 @@
+	searcher search.Searcher
+}
+
+func (ds *datastoreImpl) Search(ctx context.Context, q *v1.Query) ([]pkgSearch.Result, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "Search")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return nil, err
+	}
+	return ds.searcher.Search(ctx, q)
+}
+
+// Count returns the number of search results from the query
+func (ds *datastoreImpl) Count(ctx context.Context, q *v1.Query) (int, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "Count")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return 0, err
+	}
+	return ds.searcher.Count(ctx, q)
+}
+
+func (ds *datastoreImpl) SearchReportSnapshots(ctx context.Context, q *v1.Query) ([]*storage.ReportSnapshot, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "SearchReportSnapshots")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return nil, err
+	}
+	return ds.searcher.SearchReportSnapshots(ctx, q)
+}
+
+func (ds *datastoreImpl) SearchResults(ctx context.Context, q *v1.Query) ([]*v1.SearchResult, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "SearchResults")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return nil, err
+	}
+	return ds.searcher.SearchResults(ctx, q)
+}
+
+func (ds *datastoreImpl) Get(ctx context.Context, id string) (*storage.ReportSnapshot, bool, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "Get")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return nil, false, err
+	}
+	snap, found, err := ds.storage.Get(ctx, id)
+	if err != nil || !found {
+		return nil, false, err
+	}
+
+	return snap, true, nil
+}
+
+func (ds *datastoreImpl) Exists(ctx context.Context, id string) (bool, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "Exists")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return false, err
+	}
+	found, err := ds.storage.Exists(ctx, id)
+	if err != nil || !found {
+		return false, err
+	}
+	return true, nil
+}
+
+func (ds *datastoreImpl) GetMany(ctx context.Context, ids []string) ([]*storage.ReportSnapshot, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "GetMany")
+	if ok, err := workflowSAC.ReadAllowed(ctx); !ok || err != nil {
+		return nil, err
+	}
+	snaps, _, err := ds.storage.GetMany(ctx, ids)
+	if err != nil {
+		return nil, err
+	}
+	return snaps, nil
+}
+
+func (ds *datastoreImpl) AddReportSnapshot(ctx context.Context, snap *storage.ReportSnapshot) error {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), resourceType, "AddReportSnapshot")
+	if err := sac.VerifyAuthzOK(workflowSAC.WriteAllowed(ctx)); err != nil {
+		return err
+	}
+	if snap.ReportId == "" {
+		return errors.New("New report snapshot must have a non-empty report id")
+	}
+	snap.ReportId = uuid.NewV4().String()
```

Comment:
this overwrites the `snap.ReportId` we just required to be not empty.  That seems odd, if that is intentional a comment explaining would be useful.

Reply:
Good catch, that should be removed.